sssd_test_framework/roles/ipa.py

```diff
@@ -2194,3 +2227,163 @@ def delete(self) -> None:
+        """Adds one or more hosts to the rule."""
+        hosts = [hosts] if isinstance(hosts, str) else hosts
+        for host in hosts:
+            self.role.host.conn.exec(["ipa", "hbacrule-add-host", self.name, f"--hosts={host}"])
+        return self
+
+    def add_service(self, services: list[str] | str) -> IPAHbac:
+        """Adds one or more services to the rule."""
+        services = [services] if isinstance(services, str) else services
+        for svc in services:
+            self.role.host.conn.exec(["ipa", "hbacrule-add-service", self.name, f"--hbacsvcs={svc}"])
+        return self
+
+    def show(self) -> dict[str, list[str]]:
+        """Displays the HBAC rule."""
+        cmd = self.role.host.conn.exec(["ipa", "hbacrule-show", self.name, "--all"])
+        # return result.stdout_lines
+        lines = [line.strip() for line in cmd.stdout_lines if ":" in line]
+
+        return attrs_parse(lines)
+
+    def delete(self) -> None:
+        """
+        Deletes the HBAC rule.
+        """
+        self.role.host.conn.exec(["ipa", "hbacrule-del", self.name])
+
+    def enable(self) -> None:
+        """
+        Enables the HBAC rule.
+        """
+        self.role.host.conn.exec(["ipa", "hbacrule-enable", self.name])
+
+    def disable(self) -> None:
+        """
+        Disables the HBAC rule.
+        """
+        self.role.host.conn.exec(["ipa", "hbacrule-disable", self.name])
+
+    def remove_users(self, users: IPAUser | list[IPAUser]) -> None:
+        """
+        Removes users from the HBAC rule.
+        """
+        self.role.host.conn.exec(["ipa", "hbacrule-remove-user", self.name, f"--users={users}"])
+
+    def add_groups(self, groups: IPAGroup | list[IPAGroup]) -> None:
+        """
+        Adds groups to the HBAC rule.
+        """
+        self.role.host.conn.exec(["ipa", "hbacrule-add-group", self.name, f"--groups={groups}"])
+
+    def remove_groups(self, groups: IPAGroup) -> None:
+        """
+        Removes groups from the HBAC rule.
+        """
+        self.role.host.conn.exec(["ipa", "hbacrule-remove-group", self.name, f"--groups={groups}"])
+
+    def remove_hosts(self, hosts) -> None:
+        """
+        Removes hosts from the HBAC rule.
+        """
+        self.role.host.conn.exec(
+            ["ipa", "hbacrule-remove-host", self.name, f"--hosts={','.join(hosts)}"],
+        )
+
+    def remove_services(self, services) -> None:
+        """
+        Removes services from the HBAC rule.
+        """
+        self.role.host.conn.exec(["ipa", "hbacrule-remove-service", self.name, f"--services={','.join(services)}"])
+
+    def hbactest(
+        self,
+        user: str,
+        host: str,
+        service: str,
+        rule: str | None = None,
+        nodetail: bool = False,
+    ) -> dict[str, list[str]]:
+        cmd = ["ipa", "hbactest", f"--user={user}", f"--host={host}", f"--service={service}"]
+        if rule:
+            cmd.append(f"--rule={rule}")
```

Comment:
Is it required to use `--rule` option with ipa hbactest here? That tests access only against a single HBAC rule. 

I think it is better to ensure that when multiple rules are configured that hbactest shows the correct access rules, that means dropping the `--rule` option.

Reply:
Correct, I guess as long as '--rule' argument is optional it is okay.